src/Xamarin.Android.Build.Tasks/Utilities/AssemblyCompression.cs

```diff
@@ -0,0 +1,101 @@
+using System;
+using System.Buffers;
+using System.IO;
+
+using K4os.Compression.LZ4;
+
+namespace Xamarin.Android.Tasks
+{
+	class AssemblyCompression
+	{
+		public enum CompressionResult
+		{
+			Success,
+			InputTooBig,
+			EncodingFailed,
+		}
+
+		public sealed class AssemblyData
+		{
+			public string SourcePath { get; internal set; }
+			public uint DescriptorIndex { get; internal set; }
+
+			public string DestinationPath;
+			public uint SourceSize;
+			public uint DestinationSize;
+
+			public AssemblyData (string sourcePath, uint descriptorIndex)
+			{
+				SetData (sourcePath, descriptorIndex);
+			}
+
+			public void SetData (string sourcePath, uint descriptorIndex)
+			{
+				if (String.IsNullOrEmpty (sourcePath))
+					throw new ArgumentException ("must not be null or empty", nameof (sourcePath));
+				SourcePath = sourcePath;
+				DescriptorIndex = descriptorIndex;
+			}
+		}
+
+		const uint CompressedDataMagic = 0x5A4C4158; // 'XALZ', little-endian
+
+		// TODO: consider making it configurable via an MSBuild property, would be more flexible this way
+		//
+		// Arbitrary limit of the input assembly size, to clamp down on memory allocation. Our unlinked Mono.Android.dll
+		// assembly (the biggest one we have) is currently (May 2020) around 27MB, so let's bump the value to 30MB times
+		// two - it should be more than enough for most needs.
+		//public const ulong InputAssemblySizeLimit = 60 * 1024 * 1024;
+
+		static readonly ArrayPool<byte> bytePool = ArrayPool<byte>.Shared;
+
+		public static CompressionResult Compress (AssemblyData data)
+		{
+			if (data == null)
+				throw new ArgumentNullException (nameof (data));
+
+			var fi = new FileInfo (data.SourcePath);
+			if (!fi.Exists)
+				throw new InvalidOperationException ($"File '{data.SourcePath}' does not exist");
+			// if ((ulong)fi.Length > InputAssemblySizeLimit) {
+			// 	return CompressionResult.InputTooBig;
+			// }
+
+			data.DestinationPath = $"{data.SourcePath}.lz4";
+			data.SourceSize = (uint)fi.Length;
+
+			byte[] sourceBytes = null;
+			byte[] destBytes = null;
+			try {
+				sourceBytes = bytePool.Rent ((int)fi.Length);
```

Comment:
Another thought on the `checked()` front, 4GB is "too big", but what about "2GB+1"?  That would be enough for this `(int)` cast to return a negative value, which can't be good.

Presumably `bytePool.Rent()` would throw, but given that `Compress()` *could* return `CompressionResult.InputTooBig`, an "unhandled exception" might be less than ideal.